src/kernel.cpp

```diff
@@ -798,3 +798,157 @@ bool CheckStakeModifierCheckpoints(int nHeight, unsigned int nStakeModifierCheck
     return true;
 }
 
+
+
+// V8 Kernel Protocol
+// Tomas Brod 05.06.2017 (dd.mm.yyy)
+// Plug proof-of-work exploit.
+// TODO: Stake modifier is included without much understanding.
+// Without the modifier, attacker can create transactions which output will
+// stake at desired time. In other words attacker can check wheter transaction
+// output will stake in the future and create transactions accordingly.
+// Thus including modifier, even not completly researched, increases security.
+// Note: Rsa or Magnitude weight not included due to multiplication issue.
+// Note: Payment age and magnitude restrictions not included as they are not
+// important in my view and are too restrictive for honest users.
+// TODO: flags?
+// Note: Transaction hash is used here even thou ppcoin devs advised against it,
+// Gridcoin already used txhash in previous kernel, trying to brute-force
+// good tx hash is not possible as it is not known what stake modifier will be
+// after the coins mature!
+
+CBigNum CalculateStakeHashV8(
+    const CBlock &CoinBlock, const CTransaction &CoinTx,
+    unsigned CoinTxN, unsigned nTimeTx,
+    uint64_t StakeModifier,
+    const MiningCPID &BoincData)
+{
+    CDataStream ss(SER_GETHASH, 0);
+    ss << StakeModifier;
+    ss << (CoinBlock.nTime & ~STAKE_TIMESTAMP_MASK);
+    ss << CoinTx.GetHash();
+    ss << CoinTxN;
+    ss << (nTimeTx & ~STAKE_TIMESTAMP_MASK);
+    CBigNum hashProofOfStake( Hash(ss.begin(), ss.end()) );
+    return hashProofOfStake;
+}
+
+int64_t CalculateStakeWeightV8(
+    const CTransaction &CoinTx, unsigned CoinTxN,
+    const MiningCPID &BoincData)
+{
+    int64_t nValueIn = CoinTx.vout[CoinTxN].nValue;
+    nValueIn /= 1250000;
```

Comment:
Is there a reason why this is divided by 1.25 million while the V3 weight is divided by 125k? Won't that produce a tiny weight?

Reply:
The weight is applied to the target, and if we go above the target we fail? Does that mean that it's 10 times harder to PoS stake with V8?

Reply:
No. It is 10 times harder to stake the first 1 - 5 version 8 blocks. Difficulty will re-adjust, as we seen on testnet.

Reply:
@denravonska I am thinking about this constant for a while. If I remember correctly other PoS coins don't have it. My understanding is that eventually the target will change accordingly to keep the block time at 90 seconds. But I wonder why it is there if it does nothing.

Reply:
It prevents the BlockTrust and ChainTrust from growing to big. And imposes lover limit on stakable utxo value.